### ISSUE TITLE
feature output difference

### ISSUE THREAD

**Issue by eldrin** [keunwoochoi/transfer_learning_music#2]
I found that the feature extracted from the easy_feature_extraction.py and the feature files in data_feat/ are not same. (Tested with GTZAN dataset)

Playing with a couple of days, I found that the running average / std are initial value. 

`In [12]: hf['ConvBNEluDr']['batchnormalization_1_running_mean'][:]
Out[12]: 
array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.], dtype=float32)

In [13]: hf['ConvBNEluDr']['batchnormalization_1_running_std'][:]
Out[13]: 
array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.], dtype=float32)
`
Also for all other weights. Could that be the reason for that?

**Comment by keunwoochoi** [comment]
I'm not completely sure but probably not. `running_whatever` sounds like something only used during training.

**Comment by eldrin** [comment]
I just checked [the paper](https://screenshots.firefox.com/gBoNUYJ6NrckHIiE/arxiv.org) and the source in Keras, and it seems that those parameters are used in the [inference phase](https://github.com/fchollet/keras/blob/cf2623b19a65d128b26c400138c291fbc98fb48d/keras/layers/normalization.py#L162) also.
 
Did you extracted the feature that are on the repo with using easy_extraction and weight on the repo? If you extracted them right after the training... maybe you've used `running_whatever`s I guess.

**Comment by eldrin** [comment]
Oh, ignore above. After close look I found that it only uses the current statistics for the batch (no update for 'moving' statistics). My bad.

**Comment by eldrin** [comment]
After playing around a bit, I am more inclining more to the BN as the main villain here.

I added helper functions to extract features in batch (m=5) and tried evaluation with GTZAN and SVC (and StandardScaler) in different settings as below:

1. keep the file order sorted by genre and extract in batch and shuffle split when do the CV
- achieved around **88%** on 10 fold CV (with GridSearchCV)
- With this setting, the samples from same extraction batch can be distributed both train / test.

2. shuffle the file order when batched extraction and fixed split at 10 fold CV
- achieved around **80%**
- With this setting, samples still shuffled in CV but same extraction batch only belong either train or test

3. m=1 (current easy_extraction.py)
 - achieved around **82%**

Having above results, I was wondering how originally the feature file in the repository extracted in terms of batch size condition..

**Comment by keunwoochoi** [comment]
Ok, that doesn't sound good. 

- In my experiment for the paper, I had a csv file that has all files sorted by first, genre names (blues - classical - ..), then track idx within each genre.
- I used a batch size of 256 with the same `predict` function in my notebook.

**Comment by keunwoochoi** [comment]
Interesting that both of my exp and yours imply that it -- the batch norm -- works better with batches filled with similar songs.

**Comment by eldrin** [comment]
Interestingly, both batch size and the extraction batch order seems to affect the performance. I just did two more experiments with following settings:

1. genre order / batch size m=50
- achieved around **96%**(!!) (88% for m=5)

2. shuffled order / batch size m=50
- achieved around **83%** (80% for m=5)

It seems batch statistics does affect the extracted features distribution, which makes them sharing some factors in features so that the train-test split is not clearly exclusive (probably harming the iid condition of observations?)

However, features coming from similar tracks' batch stats are more strong (even considering above possible problem), which is still interesting.

Another interesting thing is actually the feature extracted in case 1 does not seem to very separable in terms of labels..

![gtzan_batch50_tsne_scatter](https://user-images.githubusercontent.com/5870216/33350944-bc724126-d4a1-11e7-9932-81b774b4df8e.png)

This is 2d tSNE embedding of the feature matrix, which achieved 96%.

Here's the [feature](https://drive.google.com/file/d/1ygTJYKao4t4cGSMDy7ZxmPDFC8NOCJY4/view?usp=sharing) and the [label](https://drive.google.com/file/d/1Ozla8dOFbv3PZPptFpIFg2xVm0YYIPio/view?usp=sharing)

**Comment by keunwoochoi** [comment]
Okay, we need a stable way to extract the feature -- and sadly, my experiments seem to be based on a less stable way (by using test batch statistics, but with bs=256 so it seemed fine). 

## What should've been done

is obviously to use running mean and std from the training phase like [keras 2.x implementation](https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py#L138). I don't think explicit broadcasting is relevant here, so what we need is a simple one-liner:

```python
 return K.batch_normalization(
                    inputs,
                    self.moving_mean,
                    self.moving_variance,
                    self.beta,
                    self.gamma,
                    epsilon=self.epsilon)
```
## Solution
It can be done using [Keras 1.2.2 backend](https://faroit.github.io/keras-docs/1.2.2/backend/) as well as below. 
```
K.batch_normalization(x, mean, var, beta, gamma, epsilon=0.001)
```

@eldrin Please have a look, hopefully it fixes your problem. I'd love to get my code fixed and PR is always welcomed :)

**Comment by keunwoochoi** [comment]
## Notes
- Keras 2.0 stores moving variance
- Keras 1.x stores moving std
- In both, the backend take **var**

For some reason I can't find the `batch_normalization` operation in keras 1 backend, but it has it.

```
>>> K.batch_normalization
<function batch_normalization at 0x34c1aa0>
>>>
>>> import keras
>>> keras.__version__
'1.1.0'
>>> K.batch_normalization.__doc__
'Apply batch normalization on x given mean, var, beta and gamma.\n    '
```

**Comment by keunwoochoi** [comment]
Have a look on this gist:
https://gist.github.com/keunwoochoi/ab3d8e17ebce8d8f881898d6ce0b746a

**Comment by eldrin** [comment]
Thanks for the kind suggestion. Can you

**Comment by eldrin** [comment]
After a quick look in the weight file,

Unfortunately I find that it doesn't contain BN running stats..

![screen shot 2017-11-30 at 17 42 28](https://user-images.githubusercontent.com/5870216/33442654-f72f8792-d5f5-11e7-8588-3d610dedd126.png)

It seems they are not saved accidentally or because of the version of Keras 1.X didn't care it about..

would it exist any other weight files you have locally?

**Comment by keunwoochoi** [comment]
Oh.. that's embarrassing. Okay, now I remembered, although it should be with `mode=0` which would do the job well, I had to use `mode=2` for some reason. -- and that'd be why doesn't have relevant info.

Now i really feel like I wanna re-train it on Keras 2.

**Comment by keunwoochoi** [comment]
https://github.com/keunwoochoi/transfer_learning_music/tree/master/keras2_model Just pushed keras2 model, the basic problem (with changing batch size) is solved. Not tested on any target task.

**Comment by eldrin** [comment]
Hmm.. I tested the new Net and found on some datasets it's still similar to the previous version (with batch size=1).

I extracted/tested the Keras 2.x new feature with [these codes](https://gist.github.com/eldrin/e0ed41e608b0c910128a38062e3894d4) (please check) and tested:

Dataset | Avg.Performance
--- | ---
Ballroom (not extended!) | 73.80%
GTZAN | 78.89%
MusicEmotionValence | 0.4633
MusicEmotionArousal | 0.6357

(I used all the feature dimensions for all datasets, and always used the `StandardScaler` in the `Pipeline`)

I ran 5 times of `easy_test.py` with N==10, which means 50 tests on random splits, and took the average performance.

It seems the classification is more vulnerable to the unstable CV condition. (regression tasks seems not much influenced.)

The extracted feature-label data can be found [here](https://drive.google.com/open?id=168U7uEO7vyJnn2NSRqUggmB95t_SiTqU).

**Comment by keunwoochoi** [comment]
First of all, they don't vary by the batch size, right? So 

> it's still similar to the previous..

By 'it' you mean the performance?

**Comment by eldrin** [comment]
Yes. It was about the performance.

And I didn't check the variance on the batch-size, since your experiment from `feature_extractor.ipynb` seemed reasonable..

**Comment by keunwoochoi** [comment]
Right. The performances on emotion prediction look alright. But it's not too strange that the scores are diff from my original experience - the batch normalization affects obviously. 79% is not so high though :(

**Comment by jordipons** [comment]
I think this discussion is particularly relevant for the random convnet features results. Have you run these experiments without setting mode=2 in batch_norm layers? Any intuition if the results (with random convnet features) would be affected?

Thanks!
